### PR TITLE
[luci/import] Add an error handler to ImporterEx

### PR DESCRIPTION
**Pull request by tomdol**
This commit adds the ability to add external error handlers to the ImporterEx.

By default the class logs all messages from potential errors occurring during the call to importVerifyModule(). Users can inject their own error handler in a form of a callable satisfying the following signature:
    void(const std::exception&)
The error handler should be injected using a new ImporterEx constructor.

Draft: https://github.com/Samsung/ONE/pull/14792

ONE-DCO-1.0-Signed-off-by: Tomasz Dolbniak <t.dolbniak@partner.samsung.com>

### CODE REVIEW

File: compiler/luci/import/include/luci/ImporterEx.h [hunk 1]
```diff
@@ -34,14 +35,12 @@ namespace luci
...
+  ImporterEx();
+  ImporterEx(const std::function<void(const std::exception &)> &error_handler);
 
 public:
   // TODO remove this after embedded-import-value-test has moved to onert-micro
-  explicit ImporterEx(const GraphBuilderSource *source) : _source{source}
-  {
-    // DO NOTHING
-  }
+  explicit ImporterEx(const GraphBuilderSource *source);
```
[the remark this answers is not in the document]
Comment (tomdol) [reply compiler/luci/import/include/luci/ImporterEx.h:43]: I've used a delegating constructor to call the default one from this one. This way the error handler is set for the version that accepts `GraphBuilderSource` too.

[the remark this answers is not in the document]
Comment (seanshpark) [reply compiler/luci/import/include/luci/ImporterEx.h:43]: OK got it. this change wasn't in #14792 and made confusion reading the change, and is hard to catch the 'why' even with your explanation.

File: compiler/luci/import/src/ImporterEx.cpp [hunk 2]
```diff
@@ -19,7 +19,6 @@
 
 #include <foder/FileLoader.h>
 
-#include <memory>
```
Comment (seanshpark) [inline compiler/luci/import/src/ImporterEx.cpp:22]: plz move this change to another PR

Comment (seanshpark) [reply compiler/luci/import/src/ImporterEx.cpp:22]: I think I intentionally added this include although it exists in header.
plz do not remove this.

File: compiler/luci/import/src/ImporterEx.cpp [hunk 3]
```diff
@@ -33,6 +32,22 @@ inline constexpr uint64_t FLATBUFFERS_SIZE_MAX = 2147483648UL; // 2GB
...
+  : _error_handler{[](const std::exception &e) { std::cerr << e.what() << std::endl; }}
+{
+}
+
+ImporterEx::ImporterEx(const std::function<void(const std::exception &)> &error_handler)
+  : _error_handler{error_handler}
+{
+  if (!error_handler)
+  {
+    throw std::runtime_error{"The error handler passed to ImporterEx is invalid"};
```
Comment (seanshpark) [inline compiler/luci/import/src/ImporterEx.cpp:45]: plz add header for `std::runtime_error`

File: compiler/luci/import/src/ImporterEx.test.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  bool error_handler_called = false;
+  ErrorHandler handler{error_handler_called};
+
+  luci::ImporterEx importer{handler};
+  importer.importVerifyModule("/non/existent.path");
+
+  ASSERT_TRUE(error_handler_called);
+}
+
+TEST(ImporterEx, throws_with_empty_handler_N)
```
Comment (seanshpark) [inline compiler/luci/import/src/ImporterEx.test.cpp:58]:
```suggestion
TEST(ImporterEx, throws_with_empty_handler_NEG)
```

File: compiler/luci/import/src/ImporterEx.test.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+{
+  std::function<void(const std::exception &)> empty_handler; // this doesn't contain any callable
+  ASSERT_THROW(luci::ImporterEx{empty_handler}, std::runtime_error);
+}
```
Comment (seanshpark) [inline compiler/luci/import/src/ImporterEx.test.cpp:62]: plz add more negative tests so that count(postive) <= count(negative)

Comment (tomdol) [reply compiler/luci/import/src/ImporterEx.test.cpp:62]: In fact we can treat all of those tests as negative since they are all about handling errors. I've added the _NEG suffixes to all of them.

File: compiler/luci/import/src/ImporterEx.cpp [hunk 6]
```diff
@@ -19,8 +19,9 @@
 
 #include <foder/FileLoader.h>
 
-#include <memory>
 #include <iostream>
+#include <memory>
```
Comment (seanshpark) [inline compiler/luci/import/src/ImporterEx.cpp:23]: plz do not relocate headers.

Comment (tomdol) [reply compiler/luci/import/src/ImporterEx.cpp:22]: Ok, I moved it back but this was just to have them in alphabetical order.

File: compiler/luci/import/src/ImporterEx.test.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/ImporterEx.h"
+
+#include <gtest/gtest.h>
+#include <sstream>
+
+TEST(ImporterEx, uses_default_error_handler_NEG)
```
Comment (seanshpark) [inline compiler/luci/import/src/ImporterEx.test.cpp:22]: about changed to NEG, figuring out how this is negative...

Q) what is this test doing? can't catch by the code.
it would help adding some comments.

File: compiler/luci/import/src/ImporterEx.test.cpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  importer.importVerifyModule("/non/existent.path");
+
+  std::cerr.rdbuf(cerr_original_buffer);
+
+  ASSERT_GT(cerr_substitute.str().length(), 0);
+}
+
+TEST(ImporterEx, calls_external_error_handler_NEG)
+{
+
```
Comment (seanshpark) [inline compiler/luci/import/src/ImporterEx.test.cpp:39]:
```suggestion
```

File: compiler/luci/import/src/ImporterEx.test.cpp (hunk 7 above)
Comment (tomdol) [reply compiler/luci/import/src/ImporterEx.test.cpp:22]: This is a negative test in a way that it fails to import a model from an invalid path. At the same time it does use the default error handler.

File: compiler/luci/import/src/ImporterEx.test.cpp (hunk 8 above)
Comment (tomdol) [reply compiler/luci/import/src/ImporterEx.test.cpp:39]: This is an empty line added by the formatter. If I remove it the code formatting CI check fails.

Comment (seanshpark) [reply compiler/luci/import/src/ImporterEx.test.cpp:42]: this is strange. @hseok-oh , is this intended?

I assume this empty line comes from following `struct ErrorHandler`
can you please move `ErrorHandler` outside of this function ?

File: compiler/luci/import/src/ImporterEx.test.cpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  luci::ImporterEx importer;
+  importer.importVerifyModule("/non/existent.path");
+
+  std::cerr.rdbuf(cerr_original_buffer);
+
+  ASSERT_GT(cerr_substitute.str().length(), 0);
+}
+
+TEST(ImporterEx, calls_external_error_handler_NEG)
```
[the remark this answers is not in the document]
Comment (dahlinPL) [reply compiler/luci/import/src/ImporterEx.test.cpp:40]: > Similarily to the one above it fails to import a non-existent model while delegating the error handling to an externally defined struct.

can you please add some comment for tests with explanation?

File: compiler/luci/import/src/ImporterEx.test.cpp (hunk 8 above)
Comment (tomdol) [reply compiler/luci/import/src/ImporterEx.test.cpp:42]: Ok, I've re-ran the formatter after removing this line and this time it left it as is. I'm almost certain this line was added by the formatter previously though.

In this case I'd rather leave this struct inside the test because it provides scoping and the struct is intended to be used only in this particular test.

Comment (seanshpark) [reply compiler/luci/import/src/ImporterEx.test.cpp:42]: ok thanks for double checking

File: compiler/luci/import/src/ImporterEx.test.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (tomdol) [reply compiler/luci/import/src/ImporterEx.test.cpp:40]: I've added the explanation to the tests arount the assertions

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Review by jinevening — APPROVED** [review]
LGTM

**Comment by jinevening** [comment]
FYI, I removed @parjong from the reviewer list, becasue he does not review this repo. If you have a special reason to get a review from him, please add him to the reviewer list again.

**Comment by tomdol** [comment]
@jinevening ok, thank you, there's no need to ask for his review.

**Comment by tomdol** [comment]
@seanshpark I think we don't need to wait for @dahlinPL's review if 2 approvals are enough to merge

**Review by dahlinPL — COMMENTED** [review]
I don't have permissions to approve, but looks good to me 👍

**Comment by jinevening** [comment]
@dahlinPL Could you share the list of the members who need review permissions?

**Comment by dahlinPL** [comment]
> @dahlinPL Could you share the list of the members who need review permissions?

i believe it will be only me, all other team members already have permissions :)

**Comment by lemmaa** [comment]
> > @dahlinPL Could you share the list of the members who need review permissions?
> 
> i believe it will be only me, all other team members already have permissions :)

@dahLinPL , I invite you to become a member of the @samsung/one team.

**Comment by lemmaa** [comment]
> i believe it will be only me, all other team members already have permissions :)

@dahLinPL , I invited you to the @samsung/one team.
